test/lint/lint-shell.py

```diff
@@ -0,0 +1,92 @@
+#!/usr/bin/env python3
+#
+# Copyright (c) 2018-2021 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+"""
+Check for shellcheck warnings in shell scripts.
+"""
+
+import subprocess, re, sys
+
+# Disabled warnings:
+DISABLED = [
+    'SC2162', # read without -r will mangle backslashes.
+]
+
+def check_shellcheck_install():
+    try:
+        subprocess.check_output(['shellcheck', '--version'])
+    except FileNotFoundError:
+        print('Skipping shell linting since shellcheck is not installed.')
+        sys.exit(0)
+
+def get_files(command):
+    output = subprocess.run(command, stdout=subprocess.PIPE, universal_newlines=True)
+    files = output.stdout.split('\n')
+
+    # remove whitespace element
+    files = list(filter(None, files))
+    return files
+
+def main():
+    check_shellcheck_install()
+
+    # build the `exclude` flag
+    exclude = '--exclude=' + ','.join(DISABLED)
+
+    # build the `sourced files` list
+    sourced_files_cmd = [
+        'git',
+        'grep',
+        '-El',
+        r'^# shellcheck shell=',
+    ]
+    sourced_files = get_files(sourced_files_cmd)
+
+    # build the `guix files` list
+    guix_files_cmd = [
+        'git',
+        'grep',
+        '-El',
+        r'^#!\/usr\/bin\/env bash',
+        '--',
+        'contrib/guix',
+        'contrib/shell',
+    ]
+    guix_files = get_files(guix_files_cmd)
+
+    # build the other script files list
+    files_cmd = [
+        'git',
+        'ls-files',
+        '--',
+        '*.sh',
+    ]
+    files = get_files(files_cmd)
+    # remove everything that doesn't match this regex
+    reg = re.compile(r'src/[leveldb,secp256k1,minisketch,univalue]')
+    files[:] = [file for file in files if not reg.match(file)]
+
+    # build the `shellcheck` command
+    shellcheck_cmd = [
+        'shellcheck',
+        '--external-sources',
+        '--check-sourced',
+        '--source-path=SCRIPTDIR',
+    ]
+    shellcheck_cmd.append(exclude)
+    shellcheck_cmd.extend(sourced_files)
+    shellcheck_cmd.extend(guix_files)
+    shellcheck_cmd.extend(files)
+
+    # run the `shellcheck` command
+    try:
+        subprocess.check_output(shellcheck_cmd, stderr=subprocess.STDOUT)
```

Comment:
No need to use check_output here, as we want to pass through the output.
Could use, say, subprocess.run with `check=True`

Reply:
Thanks for the review. :)

I believe this might introduce slightly different behavior with output from the old script.

With the old script there would be no output if the shellcheck tests pass, but while using `subprocess.run(shellcheck_cmd, check=True)`, I see the following if the tests pass:

```
$ ./test/lint/lint-shell.py 
ShellCheck - shell script analysis tool
version: 0.8.0
license: GNU General Public License, version 3
website: https://www.shellcheck.net
```

The exit codes still work the same, but are there downstream impacts with this standard output change?

Reply:
@whiteh0rse Good point. 
One way to suppress this unnecessary stdout while using `subprocess.run` is to add `stdout = subprocess.DEVNULL` argument.

Reply:
I figured out this output from shellcheck wasn't coming from this line. Rather, it was coming from `check_shellcheck_install()` function. I piped the output from that command to `subprocess.DEVNULL`. Thanks for the suggestion, @Eunoia1729.